### PR TITLE
Feature flag: Elastic IPs for Apps

### LINKED ISSUE #50: Elastic IP support
Without ELB's it would be fairly difficult to continuously reach a certain set of instances. Having EIP support would make this a lot easier.

(also output the EIP(s) after the deploy is done)

### PR DESCRIPTION
**Pull request by DeviaVir**
Adds feature flag to apps: `elastic_ips: "enabled||disabled"`

Fixes #50

This is still missing:
  - [x] ~~test integration~~
  - [x] ~~the [ingress_resource suggestion](https://github.com/pebble/spacel-provision/issues/50#issuecomment-260122361)~~

### CODE REVIEW

File: src/spacel/provision/template/app.py [hunk 1]
```diff
@@ -86,6 +87,54 @@ def app(self, app, region):
...
+                'Properties': {
+                    'PolicyName': 'AssociateElasticIpAddress',
+                    'Roles': [{'Ref': 'Role'}],
+                    'PolicyDocument': {
+                        'Statement': [
+                            {
+                                'Effect': 'Allow',
+                                'Action': [
+                                    'ec2:AssociateAddress',
+                                    'ec2:DescribeAddresses'
```
Comment (DeviaVir) [inline src/spacel/provision/template/app.py]: Now if an EIP is already associated, a new instance (rolling update) does not have the permissions to describe and disassociate (it's broken), do we add these permissions, or is there a way around that I haven't considered yet?

**Comment by DeviaVir** [comment]
```
.
----------------------------------------------------------------------
Ran 1 test in 7.882s

OK
```

File: src/spacel/provision/template/app.py (hunk 1 above)
Comment (thepwagner) [inline src/spacel/provision/template/app.py]: I think so; it _should_ steal the EIP from instances with the older LaunchConfiguration (and rolling updates work?)

File: src/spacel/provision/app/ingress_resource.py [hunk 2]
```diff
@@ -102,3 +111,19 @@ def app_sg(self, orbit, region, app, sg_ref='Sg'):
                              orbit.name, region)
                 return None
             raise e
+
+    def app_eips(self, orbit, region, app, eips=[]):
```
Comment (thepwagner) [inline src/spacel/provision/app/ingress_resource.py]: Nit: `_app_eips()` to discourage public use (this smells like a helper?)

Comment (thepwagner) [inline src/spacel/provision/app/ingress_resource.py]: `eips=[]`: mutable arguments are evil.

I think if you run this function twice, the second execution will return duplicate elements.

File: src/spacel/provision/app/ingress_resource.py [hunk 3]
```diff
@@ -102,3 +111,19 @@ def app_sg(self, orbit, region, app, sg_ref='Sg'):
...
+        cloudformation = self._clients.cloudformation(region)
+        stack_name = '%s-%s' % (orbit.name, app)
+        try:
+            index = len(eips) + 1
+            resource = cloudformation.describe_stack_resource(
```
Comment (thepwagner) [inline src/spacel/provision/app/ingress_resource.py]: Calling `.describe_stack_resources()` once feels better than calling `.describe_stack_resource()` until it throws an error; 1 vs N server trips.

http://boto3.readthedocs.io/en/latest/reference/services/cloudformation.html#CloudFormation.Paginator.ListStackResources looks like the ultimate answer? no limits

File: src/spacel/provision/template/app.py [hunk 4]
```diff
@@ -213,12 +262,28 @@ def app(self, app, region):
...
             del params['PrivateElbSubnet01'], params['PublicElbSubnet01']
             del (resources['PublicElb'], resources['PrivateElb'],
-                 resources['ElbSg'], resources['DnsRecord'],
+                 resources['ElbSg'],
                  resources['ElbHealthPolicy'],
                  resources['Asg']['Properties']['LoadBalancerNames'])
             resources['Asg']['Properties']['HealthCheckType'] = 'EC2'
             params['ElbScheme']['Default'] = 'disabled'
 
+            # Resolve the DNS record
```
Comment (thepwagner) [inline src/spacel/provision/template/app.py]: Noice!

We _could_ tie these entries to N `Route53::HealthCheck`s too? Some day...

File: src/spacel/provision/app/ingress_resource.py (hunk 2 above)
Comment (DeviaVir) [inline src/spacel/provision/app/ingress_resource.py]: I keep doing this.. oh man. Is this the right way?

```
def app_eips(..., eips=None):
  [cont]
```

Comment (thepwagner) [inline src/spacel/provision/app/ingress_resource.py]: Yep; that'd work (with `eips = eips or []` in the function body).

Or don't make it a default value: having the non-recursive call provide a `[]` itself would also work.

File: src/spacel/provision/template/app.py (hunk 4 above)
Comment (DeviaVir) [inline src/spacel/provision/template/app.py]: For sure. Do note that those are not free (https://aws.amazon.com/route53/pricing/) and would likely have to be opt-in, but will breadcrumb.

Would be splendid if we could have a multi region weighted DNS record automatically set up...

**Comment by DeviaVir** [comment]
@thepwagner re-checked with `test integ` too, could you check again?

File: src/test/provision/app/test_ingress_resource.py [hunk 5]
```diff
@@ -39,22 +40,31 @@ def test_ingress_resources_region(self):
...
         resources = self._http_ingress(OTHER_REGION)
         # 1 per Nat EIP:
         self.assertEquals(3, len(resources))
         for resource in resources.values():
             self.assertIn('CidrIp', resource['Properties'])
 
+    def test_ingress_resources_other_region(self):
+        self.ingress._app_eips = MagicMock(return_value=['1.1.1.1', '2.2.2.2'])
+        resources = self._http_ingress(OTHER_REGION)
+        # 1 per Nat EIP:
```
Comment (thepwagner) [inline src/test/provision/app/test_ingress_resource.py]: Nit: comment lies (these aren't NAT eips?)

File: src/test/provision/app/test_ingress_resource.py [hunk 6]
```diff
@@ -65,19 +75,51 @@ def test_app_sg(self):
...
         self.assertIsNone(sg)
 
     def test_app_sg_error(self):
         self._describe_stack_resource_error('Kaboom')
-        self.assertRaises(ClientError, self.ingress.app_sg, self.orbit, REGION,
+        self.assertRaises(ClientError, self.ingress._app_sg, self.orbit, REGION,
                           'test-app')
 
+    def test_app_eips(self):
+        resourceList = [{
```
Comment (thepwagner) [inline src/test/provision/app/test_ingress_resource.py]: Pep8: `resource_list`. Meh.

**Pull request merged**